std/bigint.d

```diff
@@ -1025,6 +1025,220 @@ public:
+        {
+            value = [ 0x01u ];
+            return;
+        }
+
+        version (BigEndian)
+        {
+            // This loop adds one to an arbitrary length array of bytes.
+            for (size_t i = (value.length - 1); i < 0u; i--)
+            {
+                if (value[i] != 0xFFu)
+                {
+                    value[i]++;
+                    break;
+                }
+
+                value[i] = 0x00u;
+
+                // If the array of bytes is maxed out, append a next byte set to one.
+                if (i == (value.length - 1))
+                {
+                    value = (0x01u ~ value);
+                    return;
+                }
+            }
+        }
+        else version (LittleEndian)
+        {
+            // This loop adds one to an arbitrary length array of bytes.
+            for (size_t i = 0u; i < value.length; i++)
+            {
+                if (value[i] != 0xFFu)
+                {
+                    value[i]++;
+                    break;
+                }
+
+                value[i] = 0x00u;
+
+                // If the array of bytes is maxed out, append a next byte set to one.
+                if (i == (value.length - 1))
+                {
+                    value ~= 0x01u;
+                    return;
+                }
+            }
+        }
+    }
+
+    @safe
+    unittest
+    {
+        ubyte[] a = [ 0x00u ];
+        incrementBytes(a);
+        assert(a == [ 0x01u ]);
+
+        ubyte[] b = [ 0xFFu ];
+        incrementBytes(b);
+        assert(b == [ 0x00u, 0x01u ]);
+
+        ubyte[] c = [ 0xFDu, 0xFFu, 0xFFu ];
+        incrementBytes(c);
+        assert(c == [ 0xFEu, 0xFFu, 0xFFu ]);
+    }
+
+    /**
+        Converts the `BigInt` to a sequence of bytes that represents the
+        native-endian representation of that number.
+    */
+    ubyte[] toBytes() const @system @property nothrow
+    in
+    {
+        assert(this.uintLength > 0u);
+    }
+    out (ret)
+    {
+        assert(ret.length > 0u);
+    }
+    body
+    {
+        ubyte[] ret;
+        ret.length = (this.uintLength << 2); // Multiply by 4
+        size_t u = 0u;
+        while (u < this.uintLength)
+        {
+            *cast(uint*) &(ret[(u << 2)]) = (this.sign ? ~this.data.peekUint(u) : this.data.peekUint(u));
+            u++;
+        }
+
+        if (this.sign) BigInt.incrementBytes(ret);
+
+        version (BigEndian)
+        {
+            size_t startOfNonPadding = 0u;
+            if (this >= 0)
+            {
+                for (size_t i = 0u; i < (ret.length - 1); i++)
+                {
+                    if (ret[i] != 0x00u) break;
+                    if (!(ret[i+1] & 0x80u)) startOfNonPadding++;
+                }
+                ret = ret[startOfNonPadding .. $];
+                if (ret[$-1] & 0x80u) ret = (0x00u ~ ret);
+            }
+            else
+            {
+                for (size_t i = 0u; i < (ret.length - 1); i++)
```

Comment:
style: there's no ambiguity with precedence so ret.length-1 can live without parens.

Reply:
by the way `foreach(i; 0.. N)` tends to be used more nowadays in D.

Reply:
That's a good tip; definitely a cleaner way to do read-only loops. However, that won't work when you have to set the values in an array, because `i` in your example would be passed in by value, so setting `i` would not affect the original array from which the slice was taken. (And indeed, I have to do that in my code, particularly when converting an array of bytes to two's complement form.)

I don't know how slicing works (base+bound vs. big dumb copy vs. copy-on-write), but if it is the latter two of those three, then changes to the slice will not modify the original as well. But I'm not qualified to assert anything else. 😝